### PR TITLE
Packaged +U into derived type

### PR DESCRIPTION
**Pull request by bhourahine**
Instead of passing around various data for +U calculations as separate variables, packages it all into a derived type.

### CODE REVIEW

File: prog/dftb+/lib_dftbplus/initprogram.F90 [hunk 1]
```diff
@@ -1241,7 +1224,11 @@ subroutine initProgramVariables(this, input, env)
     ! Basic variables
     this%hamiltonianType = input%ctrl%hamiltonian
     this%tSccCalc = input%ctrl%tScc
-    this%tDFTBU = input%ctrl%tDFTBU
+    if (allocated(input%ctrl%dftbU_inp)) then
+      allocate(this%dftbU)
+      call TDftbU_init(this%dftbU, input%ctrl%dftbU_inp)
+      deallocate(input%ctrl%dftbU_inp)
```
Comment (aradi) [inline prog/dftb+/lib_dftbplus/initprogram.F90:1230]: Let's leave away the `deallocate`, we are destroying input not much later anyway.

File: prog/dftb+/lib_dftb/dftbplusu.F90 [hunk 2]
```diff
@@ -16,15 +16,80 @@ module dftbp_dftbplusu
...
 
-  public :: getDftbUShift, AppendBlock_reduce, Block_expand
-  public :: E_DFTBU, DFTBplsU_getOrbitalEquiv, DFTBU_blockIndx
-  public :: plusUFunctionals
+  public :: TDftbU_inp, TDftbU, TDftbU_init, plusUFunctionals, AppendBlock_reduce
 
-  
-  !> Contains functional characteristics
+  !> Input for DFTB+U calculation
+  type TDftbU_inp
```
Comment (aradi) [inline prog/dftb+/lib_dftb/dftbplusu.F90:26]: I'd favour `TDftbInp` or `TDftbInput`

File: prog/dftb+/lib_dftb/dftbplusu.F90 [hunk 3]
```diff
@@ -16,15 +16,80 @@ module dftbp_dftbplusu
...
+    procedure, private :: shift_U
+
+    !> Complex shift from +U potentials
+    procedure, private :: shift_iU
+
+    !> Shift from +U potentials
+    generic :: getDftbUShift => shift_U, shift_iU
+
+    !> Expand charges from reduced vector
+    procedure :: Block_expand
```
Comment (aradi) [inline prog/dftb+/lib_dftb/dftbplusu.F90:75]: What about `expandBlock`?

File: prog/dftb+/lib_dftb/dftbplusu.F90 [hunk 4]
```diff
@@ -16,15 +16,80 @@ module dftbp_dftbplusu
...
+
+    !> Equivalence between orbitals
+    procedure :: getOrbitalEquiv
+
+    !> Generates index for packing atomic blocks into 1D array
+    procedure :: blockIndx
+
+    !> Energy from +U
+    procedure :: E_DFTBU
```
Comment (aradi) [inline prog/dftb+/lib_dftb/dftbplusu.F90:84]: What about `getEnergy`?

**Review by aradi — APPROVED** [review]
Very welcomed PR! Apart of the terminological suggestion it is ready to go.

File: prog/dftb+/lib_dftb/dftbplusu.F90 (hunk 2 above)
Comment (bhourahine) [reply prog/dftb+/lib_dftb/dftbplusu.F90:26]: How about TDftb**U**Inp ?

File: prog/dftb+/lib_dftbplus/initprogram.F90 (hunk 1 above)
Comment (bhourahine) [reply prog/dftb+/lib_dftbplus/initprogram.F90:1230]: OK, but there are various other input examples with deallocations, so shall we make this an issue for a future PR to clean up?

**Comment by bhourahine** [comment]
@aradi Just waiting on buildbot checks to finish, but also hold off until the two conversations are resolved.

**Comment by bhourahine** [comment]
> @aradi Just waiting on buildbot checks to finish, but also hold off until the two conversations are resolved.

Buildbot passes.

File: prog/dftb+/lib_dftb/dftbplusu.F90 (hunk 2 above)
Comment (aradi) [reply prog/dftb+/lib_dftb/dftbplusu.F90]: Yes, of course, it was just a typo. I meant `TDftbUInp` or `TDftbUInput'.

File: prog/dftb+/lib_dftbplus/initprogram.F90 (hunk 1 above)
Comment (aradi) [reply prog/dftb+/lib_dftbplus/initprogram.F90]: Valid point, so probably it is indeed better to leave it this way for the moment.

File: prog/dftb+/lib_dftb/blockpothelper.F90 [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  use dftbp_commontypes, only : TOrbitals
+  implicit none
+
+  private
+  public :: AppendBlock_reduce
+
+contains
+
+  !> Adds block (dual) charges onto end of a 1D vector of reduced charges
+  subroutine AppendBlock_reduce(input, equiv, orb, output, isSkew)
```
Comment (aradi) [inline prog/dftb+/lib_dftb/blockpothelper.F90:23]: Could we use the occasion of touching this, to align the subroutine name with the style guide? E.g. `appendBlockReduced` or similar?

File: prog/dftb+/lib_dftbplus/inputdata.F90 [hunk 6]
```diff
@@ -360,7 +360,7 @@ module dftbp_inputdata
     real(dp), allocatable :: xi(:,:)
 
     !> DFTB+U input, if present
-    type(TDftbU_inp), allocatable :: dftbU_inp
+    type(TDftbUInp), allocatable :: dftbU_inp
```
Comment (aradi) [inline prog/dftb+/lib_dftbplus/inputdata.F90:363]: `:: dftbUInp` as variable name?

**Review by aradi — APPROVED** [review]

File: prog/dftb+/lib_dftbplus/initprogram.F90 (hunk 1 above)
Comment (bhourahine) [reply prog/dftb+/lib_dftbplus/initprogram.F90]: As its now changed, I'll leave it this way. But the question is whether to clean up these patterns globally in another PR if initprogram is getting refactored.

**Pull request merged**